dataservice/api/genomic_file/models.py

```diff
@@ -30,9 +49,167 @@ class GenomicFile(db.Model, Base):
+    """
+    Registers the genomic file with indexd by sending a POST with simple
+    auth to /index/ with a document formatted as:
+
+      {
+        "file_name": "my_file",
+        "hashes": {
+          "md5": "0b7940593044dff8e74380476b2b27a9"
+        },
+        "size": 123,
+        "urls": [
+          "s3://mybucket/mykey/"
+        ],
+        "metadata": {
+          "acls": "phs000000",
+          "kf_id": "PT_00000000"
+        }
+      }
+
+    The response upon successful registry will contain a `did` which will
+    be used as the target's uuid so that it may be joined with the indexd
+    data.
+    """
+    if current_app.config['INDEXD_URL'] is None:
+        return
+
+    req_body = {
+        "file_name": target.file_name,
+        "size": target.size,
+        "form": "object",
+        "hashes": {
+            "md5": str(target.md5sum).replace('-', '')
+        },
+        "urls": [
+            target.file_url
+        ],
+        "metadata": target._metadata
+    }
+
+    # Register the file on indexd
+    resp = requests.post(current_app.config['INDEXD_URL'],
+                         auth=(current_app.config['INDEXD_USER'],
+                               current_app.config['INDEXD_PASS']),
+                         headers={'Content-Type': 'application/json'},
+                         json=req_body)
+
+    try:
+        resp.raise_for_status()
+    except HTTPError:
+        message = 'could not register genomic_file'
+        # Attach indexd error message, if there is one
+        if 'error' in resp.json():
+            message = '{}: {}'.format(message, resp.json()['error'])
+        abort(500, message)
+
+    resp = resp.json()
+    target.uuid = resp['did']
+    return target
+
+
+@event.listens_for(GenomicFile, 'before_update')
+def update_indexd(mapper, connection, target):
+    """
+    Updates a document in indexd
+    """
+    if current_app.config['INDEXD_URL'] is None:
+        return
+
+    # Get the current revision if not already loaded
+    if target.rev is None:
+        resp = requests.get(current_app.config['INDEXD_URL']+target.uuid)
+        resp.raise_for_status()
+        target.rev = resp.json()['rev']
+
+    req_body = {
+        "file_name": target.file_name,
+        "size": target.size,
+        "hashes": {
+            "md5": target.md5sum
+        },
+        "urls": [
+            target.file_url
+        ],
+        "metadata": target._metadata
+    }
+
+    # Update the file on indexd
+    url = '{}{}?rev={}'.format(current_app.config['INDEXD_URL'],
+                               target.uuid,
+                               target.rev)
+    resp = requests.put(url,
+                        auth=(current_app.config['INDEXD_USER'],
+                              current_app.config['INDEXD_PASS']),
+                        json=req_body)
+
+    try:
+        resp.raise_for_status()
+    except HTTPError:
+        abort(500, 'could not update genomic_file')
```

Comment:
I would like to, but that endpoint doesn't return any messages or json when it errors. I could throw something in there in hopes that it will one day return something.

Reply:
We shouldn't send back anything about gen3, user's shouldn't be aware of this concept. We should log it internally, though, but I'm not yet sure how we want to take care of that.